doc/man/man1/ocpiremote.1
.\"     Title: ocpiremote
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 08/20/2020
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "OCPIREMOTE" "1" "08/20/2020" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ocpiremote \- command\-line tool for remotely managing an OpenCPI container server
.SH "SYNOPSIS"
.sp
\fBocpiremote\fR [\fI<options>\fR] [\fI<verb>\fR]
.SH "DESCRIPTION"
.sp
The \fBocpiremote(1)\fR command\-line tool can be used on a local system to load, start, and control an \fBocpiserve(1)\fR container server on a remote system\&. The target remote system requires the Secure Shell (SSH), the \fBbash\fR shell and the \fBtar\fR command in addition to an OpenCPI\-supported software (RCC) platform\&. No additional installation is necessary\&.
.sp
The \fBocpiremote\fR tool loads an \fBocpiserve\fR executable "package" with Secure Copy Protocol (SCP) over an SSH connection into a "sandbox" directory (by default, \fB~/sandbox/\fR) on the remote system and can subsequently be used to start it\&. When started, the \fBocpiserve\fR container server makes its local containers available to any and all remote clients through the local network interfaces on the remote system\&. See the ocpiserve(1) tool for details\&. The \fBocpiremote\fR tool manages the container server in the sandbox directory without affecting the rest of the remote system\(cqs installation and provides a convenient way to set up and operate container servers remotely\&.
.sp
The \fBocpiremote\fR command\-line tool is part of OpenCPI\(cqs \fB\fIremote containers\fR\fR feature, which allows for the addition of containers available in other systems on the network to the set of containers considered for application execution\&. See the chapter "Using Remote Containers: Network\-Connected Processors" in the \fIOpenCPI Application Development Guide\fR for information on how to set up and use this feature\&.
.SH "VERBS"
.sp
The \fBocpiremote\fR command operates on the remote system whose IP address and TCP port are listed first in the OCPI_SERVER_ADDRESSES environment variable or whose IP address and (optional) TCP port are specified as options on the command line\&.
.sp
The verbs supported by \fBocpiremote\fR are:
.PP
\fBdeploy\fR
.RS 4
Copy an OpenCPI bootable media directory tree (also called a "deployment package" or "SD card contents") to a remote system (usually to an SD card inserted into that system) and then reboot it\&. See the
\fIOpenCPI Installation Guide\fR, the
\fIOpenCPI Platform Development Guide\fR, and the
ocpiadmin(1)
tool for more information about creating bootable media directory trees for embedded systems\&.
.RE
.PP
\fBload\fR
.RS 4
Create a sandbox directory on a remote system and then send an
\fBocpiserve\fR
container server package to that directory\&.
.RE
.PP
\fBlog\fR
.RS 4
Interactively view
\fBocpiserve\fR
container server logging\&.
.RE
.PP
\fBreboot\fR
.RS 4
Reboot a remote system\&.
.RE
.PP
\fBreload\fR
.RS 4
Delete an
\fBocpiserve\fR
container server sandbox directory and then reload it\&.
.RE
.PP
\fBrestart\fR
.RS 4
Stop and then start an
\fBocpiserve\fR
container server on a remote system\&.
.RE
.PP
\fBstart\fR
.RS 4
Start an
\fBocpiserve\fR
container server on a remote system\&.
.RE
.PP
\fBstatus\fR
.RS 4
Get the status of an
\fBocpiserve\fR
container server on a remote system\&.
.RE
.PP
\fBstop\fR
.RS 4
Stop an
\fBocpiserve\fR
container server on a remote system\&.
.RE
.PP
\fBtest\fR
.RS 4
Test the basic connectivity between the local system and an
\fBocpiserve\fR
container server on a remote system\&.
.RE
.PP
\fBunload\fR
.RS 4
Delete a container server sandbox directory\&.
.RE
.sp
None of these verbs take any arguments\&.
.SH "NOUNS"
.sp
There are no nouns for \fBocpiremote\fR\&.
.SH "OPTIONS"
.sp
In the descriptions below, a plus sign (+) after the option indicates that it can be specified more than once on the command line\&.
.SS "Options Common to All Operations on a Container Server or Remote System"
.PP
\fB\-\-ip_addr=\fR\fI<ip\-address>\fR, \fB\-i\fR \fI<ip\-address>\fR
.RS 4
Specify the IP address of the target remote system\&. If this option is not specified, the first IP address in the OCPI_SERVER_ADDRESSES environment variable is used\&.
.RE
.PP
\fB\-\-password=\fR\fI<password>\fR, \fB\-p\fR \fI<password>\fR
.RS 4
Specify the password associated with the username used for login to the target remote system\&. The default password is
\fBroot\fR\&.
.RE
.PP
\fB\-\-remote_dir=\fR\fI<directory\-name>\fR, \fB\-d\fR \fI<directory\-name>\fR
.RS 4
Specify the name of a directory on the target remote system in which to create and use a container server sandbox\&. If this option is not used, the default directory name
\fBsandbox/\fR
is created in the home directory of the specified user\&.
.RE
.PP
\fB\-\-ssh_opts=\fR\fI<options>\fR, \fB\-o\fR \fI<options>\fR
.RS 4
Specify the SSH options for connecting to the target remote system\&. If this option is not used, the SSH options for automatically accepting host keys without notification and for suppressing most warning and diagnostic messages are used\&.
.RE
.PP
\fB\-\-scp_opts=\fR\fI<options>\fR, \fB\-c\fR \fI<options>\fR
.RS 4
Specify the SCP options for connecting to the target remote system\&. If this option is not used, the SCP options for automatically accepting host keys without notification and for suppressing most warning and diagnostic messages are used\&.
.RE
.PP
\fB\-\-user=\fR\fI<username>\fR, \fB\-u\fR \fI<username>\fR
.RS 4
Specify the username used for login to the target remote system\&. The default username is
\fBroot\fR\&.
.RE
.SS "Options When Loading or Reloading a Container Server"
.PP
\fB\-\-\fR\fBhw_platform=\fR\fI<HDL\-platform\-identifier>\fR, \fB\-s\fR \fI<HDL\-platform\-identifier>\fR
.RS 4
Specify the identifier of the HDL (FPGA) platform in use for the container server environment on the target remote system\&. If this option is not used, the default is
\fBzed\fR\&. See the
\fIOpenCPI User Guide\fR
for the possible identifiers for OpenCPI HDL platforms\&.
.RE
.PP
\fB\-\-port=\fR\fI<tcp\-port\-number>\fR, \fB\-r\fR \fI<tcp\-port\-number>\fR
.RS 4
Specify the TCP port of the container server on the target remote system\&. If this option is not specified, the first port number in the OCPI_SERVER_ADDRESSES environment variable is used\&.
.RE
.PP
\fB\-\-sw_platform=\fR\fI<RCC\-platform\-identifier>\fR, \fB\-s\fR \fI<RCC\-platform\-identifier>\fR
.RS 4
Specify the identifier of the RCC (software) platform in use for the container server environment on the target remote system\&. If this option is not used, the default is
\fBxilinx13_4\fR\&. See the
\fIOpenCPI User Guide\fR
for the possible identifiers for OpenCPI RCC platforms\&.
.RE
.SS "Options When Deploying OpenCPI Bootable Media Files to a Remote System"
.PP
\fB\-\-hw_platform=\fR\fI<HDL\-platform\-identifier>\fR, \fB\-s\fR \fI<HDL\-platform\-identifier>\fR
.RS 4
Specify the identifier of the HDL (FPGA) platform in use for the container server environment on the target remote system\&. If this option is not used, the default is
\fBzed\fR\&. See the
\fIOpenCPI User Guide\fR
for the possible identifiers for OpenCPI HDL platforms\&.
.RE
.PP
\fB\-\-sw_platform=\fR\fI<RCC\-platform\-identifier>\fR, \fB\-s\fR \fI<RCC\-platform\-identifier>\fR
.RS 4
Specify the identifier of the RCC (software) platform in use for the container server environment on the target remote system\&. If this option is not used, the default is
\fBxilinx13_4\fR\&. See the
\fIOpenCPI User Guide\fR
for the possible identifiers for OpenCPI RCC platforms\&.
.RE
.SS "Options When Starting or Restarting a Container Server"
.PP
\fB\-\-log\-level=\fR\fI<value>\fR, \fB\-l\fR\fI<value>\fR
.RS 4
Specify the OpenCPI log level to be used during container server execution\&. See the description of the OCPI_LOG_LEVEL environment variable in the
\fIOpenCPI User Guide\fR
for a description of possible values\&. Using this option overrides the level set in the OCPI_LOG_LEVEL environment variable\&. The default log level is 0 (no logging)\&.
.RE
.SH "ENVIRONMENT VARIABLES"
.sp
The following OpenCPI environment variables affect \fBocpiremote\fR execution:
.PP
\fBOCPI_ENABLE_REMOTE_DISCOVERY\fR
.RS 4
Whether (set to
\fB1\fR) or not (set to
\fB0\fR) multicast User Datagram Protocol (UDP) discovery of remote systems running
\fBocpiserve\fR
is enabled\&.
.RE
.PP
\fBOCPI_SERVER_ADDRESSES\fR
.RS 4
A comma\- or space\-separated list of IP addresses/TCP ports of remote systems to contact, where each IP/TCP address is in the format
\fI<ip\-address>\fR\fB:\fR\fI<tcp\-port>\fR\&. For example, 192\&.168\&.1\&.102:12345\&. Starting
\fBocpiserve\fR
with the
\fB\-\-verbose\fR
option directs it to output its IP addresses/TCP ports\&.
.RE
.PP
\fBOCPI_SERVER_ADDRESSES_FILE\fR
.RS 4
The name of a file that contains remote system IP addresses/TCP ports, one per line\&. Each IP address is in the format
\fI<ip\-address>\fR\fB:\fR\fI<tcp\-port>\fR\&. For example, 192\&.168\&.1\&.102:12345\&.
.RE
.PP
\fBOCPI_OCPI_SOCKET_INTERFACE\fR
.RS 4
The name of the network interface (from the list of network interfaces returned by the Linux
\fBifconfig\fR
command) to be used for contacting remote systems running
\fBocpiserve\fR
container servers\&. Set this variable to the interface with the same network address (IP address anded with the netmask) as the container servers being used\&.
.RE
.sp
The chapter "Using Remote Containers: Network\-Connected Processors" in the \fIOpenCPI Application Development Guide\fR provides more information about when to use these environment variables and why\&.
.SH "EXAMPLES"
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
Load an
\fBocpiserve\fR
container server on the remote system with IP address 192\&.168\&.1\&.102\&. The operation creates the container server executable in the directory
\fB~/sandbox/\&.\fR
.sp
.if n \{\
.RS 4
.\}
.nf
ocpiremote \-\-ip_addr=192\&.168\&.1\&.102 load
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
Start a previously loaded
\fBocpiserve\fR

CHANGELOG.md
# [v2.0.0-rc.1](https://gitlab.com/opencpi/opencpi/-/compare/v2.0.0-beta.1...v2.0.0-rc.1) (2020-08-28)

Changes/additions since OpenCPI Release v2.0.0-beta.1

### Enhancements
- **comp**: improve dc_offset_filter doc by adding transfer function. (!333)(364bb72e)
- **comp**: build socket_write.rcc for ubuntu18_04 platform. (!353)(7ef32634)
- **devops,tests**: remove obsolete ".bitz" files from git repo. (!337)(3d801f21)
- **doc**: add more man pages for OpenCPI commands. (!347)(6b7e41eb)
- **doc,tools**: add doc RPM to existing packaging machinery. (!329)(62cc8e74)
- **hdl base**: all hdl platforms have a standard test bitstream exported in a standard place. (!338)(e19fc095)
- **runtime,tools**: a proxy can have optional slaves so apps can have subsets, and the proxy can test or presence. (!338)(e19fc095)
- **tools**: enable python2 ACI in addition to python3 ACI. (!334)(3ea0db4d)
- **tools**: implement "-v" (verbose) option in "deploy-platform.sh" script. (!345)(5d7c5383)
- **tools**: add VERSION file that defines the version of OpenCPI. (!346)(b1e80c4d)
- **tools**: refactor `scripts/update-release.sh` to use new VERSION file. (!346)(b1e80c4d)

### Bug Fixes
- **comp**: `timestamper_scdcd.vhd`: add logic to confirm that EOF is valid. (!314)(b4918d13)
- **hdl base**: `complex_short_with_metadata_demarshaller.vhd`: add reset for EOF. (!314)(b4918d13)
- **hdl base**: `time_downsampler.vhd`: add reset for EOF. (!314)(b4918d13)
- **hdl base**: the ml605 now builds everything including the inactive and tutorial projects. (!338)(e19fc095)
- **runtime**: some property setting patterns, especially for remote device slaves, were broken and now fixed. (!344)(fb8e6c12)
- **tools**: fix export logic so as not to create broken symlinks. (!329)(62cc8e74)
- **tools**: build regressions on lesser used platforms not in normal pipeline. (!331)(a7f7cdd2)
- **tools**: fix "ocpigr" error exceptions caused by obsolete bitstream files. (!337)(3d801f21)
- **tools**: ocpigen: check status code returned by `finalizeHdlDataPort()` instead of ignoring it. (!339)(33f94239)
- **tools**: fix building an ubuntu docker container with a non ubuntu host by preventing kernel related items from being built. (!343)(29a8c490)
- **tools**: fix OSP HDL platform export mechanism. (!345)(5d7c5383)
- **tools**: fix/restore usable of project-level componentlibrary settings in proxies in hdl/cards. (!350)(1584caf7)

### Miscellaneous
- **tools**: remove support for CentOS 6 as a development host platform. (!354)(3dc24ee0)

# [v2.0.0-beta.1](https://gitlab.com/opencpi/opencpi/-/compare/v1.7.0...v2.0.0-beta.1) (2020-07-24)

Changes/additions since OpenCPI Release v1.7.0

### New Features
- **devops**: upload entire OpenCPI tree to AWS S3 when a job fails. (!297)(6708dc22)
- **osp**: add support for additional xilinx software platforms: xilinx13_4_aarch32, xilinx19_2_aarch{32,64}. (!173)(5e034a9a)
- **tools**: add AV GUI to ubuntu18_04 platform. (!294)(d74f201c)
- **tools**: add centos8 RCC development platform. (!312)(c85641a3)
- **tools**: initial add of "ubuntu18_04" development platform. (!278)(e053aac5)

### Enhancements
- **devops**: add ability for opencpi pipeline to launch downstream plutosdr pipeline. (!303)(ddb5b49e)
- **devops**: add weekly scheduled CI pipeline to lessen load on MR pipelines. (!287)(c47409e0)
- **devops**: move CI artifact hosting from gitlab to AWS. (!325)(8c28c710)
- **doc**: add generation of html man pages. (!327)(196ab92e)
- **doc**: add man pages for ocpidev nouns (previous ones are for verbs). (!327)(196ab92e)
- **doc,tools**: export PDFs so they can more easily be packaged in an RPM. (!311)(fb4c4f4c)
- **hdl base**: allow bitstream loading in linux 4.10, and avoid probing a bitstream when unloading or loading. (!316)(d3dabaf9)
- **hdl base,tools**: component library searching is now all-libs-in-project, per project, not each-lib-in-all-projects. (!316)(d3dabaf9)
- **hdl base,tools**: primitive libraries can be qualified by the project's package id. (!316)(d3dabaf9)
- **hdl base,tools**: workers chosen for an assembly are not affected by what is built as long as anything is built. (!316)(d3dabaf9)
- **osp**: add initial support of ad9361 fmcomms transceiver for zcu104 platform. (!302)(bfe43f20)
- **tools**: `ocpigr`: emit yaml instead of xml to support grc 3.8. (!306)(b2b8617d)
- **tools**: refactor ocpiadmin.sh and relocate to `tools/scripts/`. (!319)(f37f3732)

### Bug Fixes
- **devops**: fix hdl platforms building in pipelines when they shouldn't. (!317)(2c557552)
- **hdl base**: check for hdl target before building zynq primitive wrapper. (!326)(308f360e)
- **tools**: fix MANPATH and PYTHONPATH settings in "opencpi-setup.sh". (!323)(3312352c)
- **tools**: fix centos7 platform installation error when swig3 package not available. (!328)(8bf294e7)
- **tools**: fix incorrect package-ids in artifacts using top-level specs (not specs in libraries). (!321)(380f6c63)
- **tools**: fix ocpigr build issues. (!324)(543ab821)
- **tools**: suppress cross-compilation for items in `build/places` that use the `-t` flag. (!307)(894f6e81)

### Miscellaneous
- **devops**: remove CentOS 6 from CI pipeline. (!308)(5006cbc1)
- **devops,tools**: revamp OpenCPI RPM package production process. (!304)(2b62a34a)
- **doc**: update ubuntu* and centos8 platform README files. (!315)(77e3a53c)
- **osp**: exclude plutosdr in assets assemblies not capable of building. (!295)(a2c8fb98)


runtime/remote-support/scripts/ocpiremote.py
#!/usr/bin/env python3
# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of OpenCPI <http://www.opencpi.org>
#
# OpenCPI is free software: you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# OpenCPI is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

from argparse import ArgumentParser
import sys
import collections
import os
import glob
import tarfile
import tempfile
import subprocess
import datetime
import _opencpi.util as ocpiutil

Option = collections.namedtuple('Option', 'short long help default required action')
Subcommand = collections.namedtuple('Subcommand', 'name func help options')
Command = collections.namedtuple('Command', 'cmd rc stderr')


def main():
    """Creates ArgumentParser and parse user args.

    Create options and commands for an argparse ArgumentParser, parse the
    user's arguments, and call the function specified by the user, passing
    the parsed arguments.
    """
    ocpi_server_addresses = os.environ.get('OCPI_SERVER_ADDRESSES')
    ip = None
    port = None

    if ocpi_server_addresses and ':' in ocpi_server_addresses:
        ip = ocpi_server_addresses.split(':')[0]
        port = ocpi_server_addresses.split(':')[1]

    option_ip = make_option(
        '-i', '--ip_addr',
        'remote server IP address; first address in OCPI_SERVER_ADDRESSES',
        default=ip,
        required=True)
    option_port = make_option(
        '-r', '--port',
        'remote server port; first port in OCPI_SERVER_ADDRESSES',
        default=port,
        required=True)
    option_user = make_option(
        '-u', '--user',
        'user name for login on remote device',
        default='root')
    option_password = make_option(
        '-p', '--password',
        'user password for login on remote device',
        default='root')
    option_ssh_opts = make_option(
        '-o', '--ssh_opts',
        'ssh options for connecting to remote device; \
            default: -q -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null',
        default='-q -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null')
    option_scp_opts = make_option(
        '-c', '--scp_opts',
        'scp options for copying files to remote device; \
            default: -q -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null',
        default='-q -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null')
    option_remote_dir = make_option(
        '-d', '--remote_dir',
        'directory on remote device to create/use as the server sandbox',
        default='sandbox')
    option_sw_platform = make_option(
        '-s', '--sw_platform',
        'software platform for server environment; default: xilinx13_4',
        default='xilinx13_4')
    option_hw_platform = make_option(
        '-w', '--hw_platform',
        'hardware platform for server environment; default: zed',
        default='zed')
    option_valgrind = make_option(
        '-v', '--valgrind',
        'load/use Valgrind',
        action='store_true')
    option_log_level = make_option(
        '-l', '--log_level',
        'specify log level',
        default=0)

    common_options = [option_user, option_password, option_ip,
                      option_remote_dir, option_ssh_opts, option_scp_opts]

    commands = []
    commands.append(make_subcommand(
        'load', load,
        'Create and send the server package to the remote sandbox directory',
        common_options
        + [option_port, option_valgrind, option_hw_platform,
           option_sw_platform]))
    commands.append(make_subcommand(
        'unload', unload,
        'delete a server sandbox directory',
        common_options))
    commands.append(make_subcommand(
        'reload', reload_,
        'delete a server sandbox directory and then reload it',
        common_options + [option_port, option_valgrind, option_hw_platform,
                          option_sw_platform]))
    commands.append(make_subcommand(
        'start', start,
        'start server on remote device',
        common_options + [option_log_level, option_valgrind]))
    commands.append(make_subcommand(
        'restart', restart,
        'stop and then start server on remote device',
        common_options + [option_log_level, option_valgrind]))
    commands.append(make_subcommand(
        'status', status,
        'get status of server on remote device',
        common_options))
    commands.append(make_subcommand(
        'stop', stop,
        'stop server on remote device',
        common_options))
    commands.append(make_subcommand(
        'test', test,
        'test basic connectivity',
        common_options))
    commands.append(make_subcommand(
        'log', log,
        'watch server log in realtime',
        common_options))
    commands.append(make_subcommand(
        'reboot', reboot,
        'reboot the remote device',
        common_options))
    commands.append(make_subcommand(
        'deploy', deploy,
        'Deploy Opencpi boot files onto the remote device and reboot',
        common_options + [option_hw_platform, option_sw_platform]))

    parser = make_parser(commands)
    args = parser.parse_args()

    # If a subcommand was passed, call it. Else print help message
    if 'func' in args:
        rc = args.func(args)

        if rc != 0:
            sys.exit(rc)
    else:
        parser.print_help()


def make_option(short, long, help, default=None, required=False, action='store'):
    """ Returns an Option NamedTuple.

    Args:
        short: short form of option flag
        long: long form of option flag
        help: message to display when -h is passed
        default: default value for option (default None)
        required: whether an option is required (default False)
        action: the action that an option performs (default 'store')

    Returns:
        Otion NamedTuple with all args set as its members
    """
    option = Option(short, long, help, default, required, action)

    return option


def make_subcommand(name, func, help, options):
    """ Returns a Subcommand NamedTuple.

    Args:
        name: what user passes to invoke subcommand
        func: function to be invoked by subcommand
        help: help message to display when -h is passed
        options: list of Option NamedTuples that correspond to subcommand

    Returns:
        Subcommand NamedTuple with all args set as its members
    """
    command = Subcommand(name, func, help, options)

    return command


def make_command(cmd, args, ocpiserver=False, ssh=True, rc=0, stderr=True):
    """ Returns a string containing a command formatted to be executed on
        remote device.

    Args:
        cmd: string containing command before formatted to execute
            on remote server
        args: parsed user arguments
        ocpiserver: if True, will format cmd to be passed to ocpiserver.sh
            (default False)
        ssh: if True, will format cmd to be passed through ssh command to
            remote device (default True)
    """
    rc = [rc] if not isinstance(rc, list) else rc

    if ocpiserver:
        cmd = 'cd "{}" && ./ocpiserver.sh "{}"'.format(args.remote_dir, cmd)
    if ssh:
        cmd = 'ssh {} {}@{} sh -c \'{}\''.format(
            args.ssh_opts, args.user, args.ip_addr, cmd)

    command = Command(cmd, rc, stderr)

    return command


def make_parser(commands):
    """ Returns an argparse ArgumentParser.

    Args:
        commands: list of Command NamedTuples to add to the parser
    """
    parser = ArgumentParser()
    subparsers = parser.add_subparsers()

    for command in commands:
        subparser = subparsers.add_parser(name=command.name, help=command.help)
        subparser.set_defaults(func=command.func)

        for option in command.options:
            if option.default is not None:
                subparser.add_argument(option.short, option.long, default=option.default,
                                       action=option.action, help=option.help)
            else:
                subparser.add_argument(option.short, option.long, required=option.required,
                                       action=option.action, help=option.help)

    return parser


def make_tar(tar_files, arcnames, tempdir):
    """ Creates a tar file and returns the path.

    Args:
        tar_files: list of files to add to the tar
        arcnames: list of file paths to achive tar_files to
        tempdir: staging directory for the tar
    """
    mode = "w:gz"
    tar_path = os.path.join(tempdir, 'tar.tgz')

    with tarfile.open(tar_path, mode,
                      format=tarfile.PAX_FORMAT, dereference=True) as tar:
        for tar_file,arcname in zip(tar_files,arcnames):
            tar.add(tar_file, arcname=arcname)

    return tar_path


def execute_commands(commands, args):
    """ Calls execute_command for each command in commands

    Args:
        commands: list of Command NamedTuples
        args: parsed user arguments
    """
    for command in commands:
        if execute_command(command, args) not in command.rc:
            sys.exit()

    return 0


def execute_command(command, args):
    """ Executes a command using subprocess.

    Args:
        command: Command NamedTuple containing data for the command to execute
        args: parsed user arguments

    Returns:
        The subprocess's return code, stdout, and stderr
    """
    with tempfile.TemporaryDirectory() as tmpdir:
        with tempfile.NamedTemporaryFile(dir=tmpdir, mode='w+b',
                                         buffering=0, delete=False) as passwd_file:
            passwd_file_path = passwd_file.name
            passwd_file.write(bytes('echo ' + args.password, 'utf-8'))
            os.chmod(passwd_file_path, 0o700)
        try:
            stderr=''
            with subprocess.Popen(
                    command.cmd.split(),
                    env={'SSH_ASKPASS': passwd_file_path, 'DISPLAY': 'DUMMY'},
                    start_new_session=True,
                    stdin=subprocess.PIPE,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    universal_newlines=True) as process:
                for stdout in process.stdout:
                    print(stdout, end='')
                if process.stderr:
                    stderr = process.stderr.read().strip()

        except Exception as e:
            print('help')
            raise ocpiutil.OCPIException(
                'SSH/SCP call failed in a way we cannot handle; quitting. {}'.format(e))
        
        if stderr and command.stderr:
            print(stderr)

        return process.returncode


def load(args):
    """ Gather files for ocpiserver package and send to remote device.

    Args:
        args: parsed user arguments
    """
    cdk = os.environ['OCPI_CDK_DIR']

    driver_list = os.path.join(cdk,args.sw_platform,'lib','driver-list')
    with open(driver_list) as driver_file:
        drivers = ['libocpi_{}_s.so'.format(driver) for line in driver_file
                   for driver in line.strip().split()]
        drivers = [os.path.join(cdk, args.sw_platform, 'lib', driver) for driver in drivers]

    localtime = os.path.join('/etc', 'localtime')
    os_driver = os.path.join(cdk, 'scripts', 'ocpi_linux_driver')
    ocpiserver = os.path.join(cdk, 'scripts', 'ocpiserver.sh')
    kernel_objects = glob.glob(os.path.join(cdk, args.sw_platform, 'lib', '*.ko'))
    rules = glob.glob(os.path.join(cdk, args.sw_platform, 'lib', '*.rules'))
    bin_files = ['ocpidriver', 'ocpiserve', 'ocpihdl', 'ocpizynq']
    bin_files = [os.path.join(cdk, args.sw_platform, 'bin', bin_file) for bin_file in bin_files]
    sdk = glob.glob(os.path.join(cdk, args.sw_platform, 'sdk', 'lib', '*'))

    # List of files to add to tar
    tar_files = (drivers
             + kernel_objects
             + rules
             + bin_files
             + sdk
             + [localtime, ocpiserver, os_driver])

    # Where to extract files to on remote device
    arcnames = [os.path.join(args.remote_dir, *tar_file[len(cdk):].split('/'))
                if tar_file.startswith(cdk) else tar_file
                for tar_file in tar_files]

    # Check for hw system.xml. If it doesn't exist, get sw system.xml
    hw_system_xml = os.path.join(cdk, 'deploy', args.hw_platform, 'opencpi', 'system.xml')
    sw_system_xml = os.path.join(cdk, args.sw_platform, 'system.xml')
    system_xml = hw_system_xml if os.path.isfile(hw_system_xml) else sw_system_xml

    tar_files.append(system_xml)
    system_xml_arc = os.path.join(args.remote_dir, *system_xml[len(cdk):].split('/'))
    arcnames.append(system_xml_arc)

    if args.valgrind:
        tar_files.append(
            os.path.join(cdk, '..', 'prerequisites', 'valgrind', args.sw_platform))
        arcnames.append(
            os.path.join(args.remote_dir, 'prerequisites', 'valgrind', args.sw_platform))

    with tempfile.TemporaryDirectory() as tempdir:
        # Prepare sandbox
        commands = []
        commands.append(make_command(
            'mkdir {}'.format(args.remote_dir),
            args))
        commands.append(make_command(
            'date -s "{}"'.format(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")),
            args))
        commands.append(make_command(
            'echo {} > {}/swplatform'.format(args.sw_platform, args.remote_dir),
            args))
        commands.append(make_command(
            'echo {} > {}/port'.format(args.port, args.remote_dir),
            args))
        commands.append(make_command(
            'ln -s scripts/ocpiserver.sh {}'.format(args.remote_dir),
            args))
        commands.append( make_command(
            'ln -s ../{} {}'.format(system_xml_arc, args.remote_dir),
            args))

        print('Preparing remote sandbox...')
        rc = execute_commands(commands, args)

        # Create server package
        print('Creating server package...')
        tar_commands = []
        tar_path = make_tar(tar_files, arcnames, tempdir)
        tar_commands.append(make_command(
            'scp {} {} {}@{}:./{}'.format(
                args.scp_opts, tar_path, args.user, args.ip_addr, args.remote_dir),
            args,
            ssh=False))
        tar_commands.append(make_command(
            'gunzip -c {}/tar.tgz | tar xf -'.format(args.remote_dir),
            args))

        print('Sending server package...')
        rc = execute_commands(tar_commands, args)

        if rc == 0:
            print('Server package sent successfully.')

    return rc


def reboot(args):
    """ Reboot the remote device.

    Args:
        args: parsed user arguments
    """
    if status(args, stderr=False) == 0:
        if stop(args) != 0:
            return 1

    command = make_command('/sbin/reboot', args)
    rc = execute_command(command, args)

    if rc in command.rc:
        print("Rebooting remote device...")

    return rc


def deploy(args):
    """ Copy boot files to the remote device and call reboot().

    Args:
        args: parsed user arguments
    """
    cdk = os.environ['OCPI_CDK_DIR']
    local_dir = '{}/{}/sdcard-{}'.format(cdk, args.hw_platform, args.sw_platform)

    if not os.path.isdir(local_dir):
        print("Error: {} does not exist".format(local_dir))
        print("Try running 'ocpiadmin deploy platform {} {}'".format(
                args.sw_platform, args.hw_platform)
            )

        return 1

    tar_files = [os.path.join(local_dir, f) for f in os.listdir(local_dir) if f != 'opencpi']

    with tempfile.TemporaryDirectory() as tempdir:
        tar_commands = []
        tar_path = make_tar(tar_files, tar_files, tempdir)

        tar_commands.append(make_command(
            'if [ ! -d {} ]; then mkdir {}; fi'.format(args.remote_dir, args.remote_dir),
            args))
        tar_commands.append(make_command(
            'scp {} {} {}@{}:./{}'.format(
                args.scp_opts, tar_path, args.user, args.ip_addr, args.remote_dir),
            args,
            ssh=False))
        tar_commands.append(make_command(
            'gunzip -c {}/tar.tgz | tar xf -'.format(args.remote_dir),
            args))

        print('Deploying Opencpi boot files to remote device from {} ...'.format(local_dir))
        rc = execute_commands(tar_commands, args)

    if rc == 0:
        print('Opencpi boot files deployed successfully.')
        rc = reboot(args)

    return rc


def unload(args):
    """ Unload ocpiserver package from remote device.

    First attempts to stop server in case it's still running.

    Args:
        args: parsed user arguments
    """
    if status(args, stderr=False) == 0:
        if stop(args) != 0:
            return 1

    command = 'rm -r {}'.format(args.remote_dir)
    command = make_command(command, args)
    rc = execute_command(command, args)

    if rc in command.rc:
        print("Server unloaded successfully.")

    return rc


def reload_(args):
    """ Reload the remote server package on the remote device by calling unload()
        followed by load().

    Args:
        args: parsed user arguments
    """
    rc = unload(args)

    if rc == 0:
        rc = load(args)

    return rc


def start(args):
    """ Start ocpiserver on remote device.

    Args:
        args: parsed user arguments
    """
    command = 'start'
    if args.valgrind:
        command += ' -V '
    if int(args.log_level) > 0:
        command += ' -l {} '.format(args.log_level)
    command = make_command(command, args, ocpiserver=True)

    rc = execute_command(command, args)

    return rc


def stop(args):
    """ Stop ocpiserver on remote device.

    Args:
        args: parsed user arguments
    """
    command = make_command('stop', args, ocpiserver=True)
    rc = execute_command(command, args)

    return rc


def restart(args):
    """ Restart the server on the remote device by calling stop() followed by start().

    Args:
        args: parsed user arguments
    """
    rc = stop(args)

    if rc == 0:
        rc = start(args)

    return rc


def status(args, stderr=True):
    """ Get the status of the ocpiserver running on remote device.

    Args:
        args: parsed user arguments
    """
    command = make_command('status', args, ocpiserver=True, stderr=stderr)
    rc = execute_command(command, args)

    return rc


def test(args):
    """ Test basic connectivity to remote device.

    Args:
        args: parsed user arguments
    """
    command = make_command('pwd', args)
    rc = execute_command(command, args)

    return rc


def log(args):
    """ Get the status of the ocpiserver running on remote device.

    Args:
        args: parsed user arguments
    """
    command = make_command('log', args, ocpiserver=True)
    rc = execute_command(command, args)

    return rc


if __name__ == '__main__':
    main()
